flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +128,293 @@ public DefaultScheduler(
+	}
+
+	// ------------------------------------------------------------------------
+	// SchedulerNG
+	// ------------------------------------------------------------------------
+
+	@Override
+	protected void startSchedulingInternal() {
+		prepareExecutionGraphForScheduling();
+		schedulingStrategy.startScheduling();
+	}
+
+	@Override
+	public boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
+		final Optional<ExecutionVertexID> executionVertexIdOptional = getExecutionVertexId(taskExecutionState.getID());
+		if (executionVertexIdOptional.isPresent()) {
+			final ExecutionVertexID executionVertexId = executionVertexIdOptional.get();
+			updateState(taskExecutionState);
+			schedulingStrategy.onExecutionStateChange(executionVertexId, taskExecutionState.getExecutionState());
+			maybeHandleTaskFailure(taskExecutionState, executionVertexId);
+			return true;
+		}
+
+		return false;
+	}
+
+	private void maybeHandleTaskFailure(final TaskExecutionState taskExecutionState, final ExecutionVertexID executionVertexId) {
+		if (taskExecutionState.getExecutionState() == ExecutionState.FAILED) {
+			final Throwable error = taskExecutionState.getError(userCodeLoader);
+			handleTaskFailure(executionVertexId, error);
+		}
+	}
+
+	private void handleTaskFailure(final ExecutionVertexID executionVertexId, final Throwable error) {
+		final FailureHandlingResult failureHandlingResult = executionFailureHandler.getFailureHandlingResult(executionVertexId, error);
+		maybeRestartTasks(failureHandlingResult);
+	}
+
+	private void maybeRestartTasks(final FailureHandlingResult failureHandlingResult) {
+		if (failureHandlingResult.canRestart()) {
+			restartTasksWithDelay(failureHandlingResult);
+		} else {
+			failJob(failureHandlingResult.getError());
+		}
+	}
+
+	private void restartTasksWithDelay(final FailureHandlingResult failureHandlingResult) {
+		final Set<ExecutionVertexID> verticesToRestart = failureHandlingResult.getVerticesToRestart();
+
+		final Set<ExecutionVertexVersion> executionVertexVersions =
+			new HashSet<>(executionVertexVersioner.recordVertexModifications(verticesToRestart).values());
+
+		final CompletableFuture<?> cancelFuture = cancelTasksAsync(verticesToRestart);
+
+		delayExecutor.schedule(
+			() -> FutureUtils.assertNoException(
+				cancelFuture.handleAsync(restartTasksOrHandleError(executionVertexVersions), getMainThreadExecutor())),
+			failureHandlingResult.getRestartDelayMS(),
+			TimeUnit.MILLISECONDS);
+	}
+
+	private BiFunction<Object, Throwable, Void> restartTasksOrHandleError(final Set<ExecutionVertexVersion> executionVertexVersions) {
+		return (Object ignored, Throwable throwable) -> {
+
+			if (throwable == null) {
+				final Set<ExecutionVertexID> verticesToRestart = executionVertexVersioner.getUnmodifiedExecutionVertices(executionVertexVersions);
+				schedulingStrategy.restartTasks(verticesToRestart);
+			} else {
+				failJob(throwable);
```

Comment:
I think this line should be removed. In legacy mode only when we fail the cancellation in `ExecutionGraph#failGlobal()`, the job should be transitioned to `FAILED`.

Reply:
But seems to me that the throwable here is also from task cancellation?

Reply:
The throwable here is from the failed task cancellation. For reference, here is what happens in legacy code when the task cancellation fails during failover: https://github.com/apache/flink/blob/a80bbf187b8a7318db3a8b117e7281eaa174c647/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FailoverRegion.java#L147-L148

From what I can see is that the execution transitions to failed and tasks are restarted if the restart strategy allows this.

Reply:
The `FailoverRegion` might not be a good reference as it is only used in `region-legacy`.
And seems to me that if cancellation fails in `FailoverRegion`, it will fail to invoke `allVerticesInTerminalState` and may leave the region in an unrecoverable state?

I cannot come up with a case that a failed cancellation can make a job unrecoverable. But I think it will be safe to keep align with `failGlobal` to fail job on cancellation failure. It should be rare case that I actually haven encountered it in production.

Reply:
Ok, we do not have to change anything immediately. However, here is some food for thought:

- If the network between JM and TM is unavailable during cancelation, the cancelation will fail, and we will transition to `FAILED`. The time period where this can happen is narrow but the scenario is not too far fetched. It should be possible to create an environment where this happens with [flink-jepsen](https://github.com/apache/flink/tree/master/flink-jepsen).
- Transitioning the job to `FAILED`, which is a globally terminal state, will stop the job execution indefintely (until it is redeployed). This is not nice operationally and not even documented.
- If we run into violations of invariants in the execution graph, the JVM should exit and be brought up again by an external process supervisor, in my opinion. Unability to cancel tasks does not violate invariants in the execution graph.
- One could retry the task cancelation indefintely until success or accept that tasks might become orphaned (but they should terminate themselves if the connection to the JobMaster is lost).

Reply:
I tried to track the cancellation future, it it `Execution#releaseFuture` and it is only completed exceptionally when `Execution#terminalStateFuture` is completed exceptionally, which should not happen from the code.
The cancellation future will wait without a timeout. But usually it completes with success cancellation, or the TM heartbeat timed out. There is a `TaskCancelerWatchDog` at TM side, which will cause the TM to exit if it fails to cancel tasks before `"task.cancellation.timeout"`. 
So seems the cancellation failure handling should not matter, since it is not expected to happen.

And agreed that transitioning the job to `FAILED` should be our last choice. We should do it only when the restart strategy says so or we are sure that the job cannot be recovered through a global/JM failover.

Reply:
> I tried to track the cancellation future, it it Execution#releaseFuture and it is only completed > exceptionally when Execution#terminalStateFuture is completed exceptionally, which should not happen from the code.

Right, I was somehow under the impression that the cancelation future fails if the cancelation operation fails.

> So seems the cancellation failure handling should not matter, since it is not expected to happen.

Then how about we remove `failJob` and add `propagateIfNonNull(throwable);`?

Reply:
Sounds good.

Reply:
Solved it slightly differently 7aa0e586d68b892fd4113865254c58897cd352ab